app/vmselect/promql/transform.go

```diff
@@ -1988,6 +1991,87 @@ func newTransformFuncSortByLabel(isDesc bool) transformFunc {
 	}
 }
 
+func newTransformFuncAlphaNumericSort(isDesc bool) transformFunc {
+	return func(tfa *transformFuncArg) ([]*timeseries, error) {
+		args := tfa.args
+		if len(args) < 2 {
+			return nil, fmt.Errorf("expecting at least 2 args; got %d args", len(args))
+		}
+		var labels []string
+		for i, arg := range args[1:] {
+			label, err := getString(arg, 1)
+			if err != nil {
+				return nil, fmt.Errorf("cannot parse label #%d for sorting: %w", i+1, err)
+			}
+			labels = append(labels, label)
+		}
+		rvs := args[0]
+		sort.SliceStable(rvs, func(i, j int) bool {
+			for _, label := range labels {
+				a := rvs[i].MetricName.GetTagValue(label)
+				b := rvs[j].MetricName.GetTagValue(label)
+				if string(a) == string(b) {
+					continue
+				}
+				if isDesc {
+					return !alphanumericLess(string(a), string(b))
+				}
+				return alphanumericLess(string(a), string(b))
+			}
+			return false
+		})
+		return rvs, nil
+	}
+}
+
+func alphanumericLess(str, nextStr string) bool {
```

Comment:
It is better form readability and performance points of view using the following algorithm for `alphanumericLess()` function:

1. Find numeric prefixes for `a` and `b`.
2. If both numeric prefixes are equal (even if they are empty), then go to 6.
3. If numeric prefix for `a` is empty, while numeric prefix for `b` isn't empty, then return `false`.
4. If numeric prefix for `a` is non-empty, while numeric prefix for `b` is empty, then return `true`.
5. If both numeric prefixes are non-empty, then return the result of `toNumber(a_numeric_prefix) < toNumber(b_numeric_prefix)`.
6. Skip equal numeric prefixes at `a` and `b` and find non-numeric prefixes on the remaining suffixes for `a` and `b`.
7. If non-numeric prefixes are equal, then skip them and go to 1.
8. Return the result of `a_nonnumeric_prefix < b_nonnumeric_prefix`.

Reply:
refactored logic